### PR TITLE
Update semicolon handling to be more like rust's

### PR DESCRIPTION
**Pull request by jfecher**
Currently, the parser expects semicolons separating each statement in a block, with an optional final semicolon at the end of a block. This is incorrect if we want to strictly follow rust's grammar however. The proper rules should be:
- Semicolons are always required after non-expression statements - even at the end of a block.
- Semicolons are optional for `if`, `for`, and `block` expressions.
- Semicolons are optional for other expressions only at the end of a block. They are required otherwise.

This PR updates chumsky so that we can use the `validate` parser to map our input from `A -> B` rather than just `A -> A`, while also emitting non-fatal errors along the way. This is used to emit semicolon errors and recover from them.

Edit: This PR now also adds the check we discussed for checking intermediate expressions in a block have the unit type. So code like `{ 1 + 2; 3 }` is now expected to fail in the type checker. I have verified all our tests and examples still build with this change.

### CODE REVIEW

File: crates/noirc_errors/src/position.rs [hunk 1]
```diff
@@ -58,11 +58,15 @@ impl<T> std::borrow::Borrow<T> for Spanned<T> {
 pub struct Span(ByteSpan);
 
 impl Span {
-    pub fn new(range: Range<u32>) -> Span {
+    pub fn new(mut range: Range<u32>) -> Span {
+        if range.start > range.end {
```
Comment (kevaundray) [inline crates/noirc_errors/src/position.rs:62]: when does this happen?

File: crates/noirc_frontend/src/hir/type_check/expr.rs [hunk 2]
```diff
@@ -222,14 +221,26 @@ pub(crate) fn type_check_expression(
             )
         }
         HirExpression::Block(block_expr) => {
-            for stmt in block_expr.statements() {
-                super::stmt::type_check(interner, stmt, errors);
-            }
+            let mut last = None;
```
Comment (kevaundray) [inline crates/noirc_frontend/src/hir/type_check/expr.rs:224]: Can we make this the unit Type instead of an optional?

Comment (jfecher) [reply crates/noirc_frontend/src/hir/type_check/expr.rs:224]: Good idea

File: crates/noirc_errors/src/position.rs [hunk 3]
```diff
@@ -58,11 +58,15 @@ impl<T> std::borrow::Borrow<T> for Spanned<T> {
...
+            range = range.end..range.start;
+        }
         Span(ByteSpan::from(range))
     }
 
     pub fn exclusive(start: u32, end: u32) -> Span {
+        println!("range at {} .. {}", start, end);
```
[the remark this answers is not in the document]
Comment (jfecher) [reply crates/noirc_errors/src/position.rs:69]: Ack, good catch

File: crates/noirc_errors/src/position.rs (hunk 1 above)
Comment (jfecher) [reply crates/noirc_errors/src/position.rs:62]: I think this is a bug in the new version of chumsky, it happens when combining ranges during parsing within some combinators now unfortunately.

Comment (kevaundray) [reply crates/noirc_errors/src/position.rs:62]: I've opened an issue for this here: https://github.com/noir-lang/noir/issues/156 so as not to block this PR

**Review by kevaundray — APPROVED** [review]
LGTM

**Pull request merged**